File: backups/20251107_125734/create_json_backups.ts
import { PrismaClient } from '@prisma/client';
import * as fs from 'fs';
import * as path from 'path';

const prisma = new PrismaClient();

interface BackupMetadata {
  tableName: string;
  recordCount: number;
  timestamp: string;
  filePath: string;
  fileSize?: number;
}

interface BackupSummary {
  backupTimestamp: string;
  databaseVersion: string;
  tables: BackupMetadata[];
  totalRecords: number;
  totalSize: number;
}

async function getDatabaseVersion(): Promise<string> {
  try {
    const result = await prisma.$queryRaw<Array<{ version: string }>>`SELECT version()`;
    return result[0]?.version || 'Unknown';
  } catch (error) {
    return 'Unknown';
  }
}

async function getFileSize(filePath: string): Promise<number> {
  try {
    const stats = fs.statSync(filePath);
    return stats.size;
  } catch (error) {
    return 0;
  }
}

async function backupTable<T>(
  tableName: string,
  fetchData: () => Promise<T[]>,
  backupDir: string
): Promise<BackupMetadata> {
  console.log(`\n📦 Backing up ${tableName}...`);

  try {
    // Fetch all records
    const records = await fetchData();
    const recordCount = records.length;

    console.log(`   Found ${recordCount} records`);

    // Create backup object with metadata
    const backup = {
      metadata: {
        tableName,
        recordCount,
        timestamp: new Date().toISOString(),
        backupVersion: '1.0.0'
      },
      data: records
    };

    // Write to file
    const fileName = `${tableName}_backup.json`;
    const filePath = path.join(backupDir, fileName);

    fs.writeFileSync(filePath, JSON.stringify(backup, null, 2), 'utf-8');

    const fileSize = await getFileSize(filePath);
    const fileSizeMB = (fileSize / 1024 / 1024).toFixed(2);

    console.log(`   ✅ Saved to ${fileName} (${fileSizeMB} MB)`);

    return {
      tableName,
      recordCount,
      timestamp: backup.metadata.timestamp,
      filePath,
      fileSize
    };
  } catch (error) {
    console.error(`   ❌ Error backing up ${tableName}:`, error);
    throw error;
  }
}

async function createBackups() {
  const backupDir = '/Users/fatihalkan/Documents/GitHub/orubacontacts/backups/20251107_125734';
  const backupMetadata: BackupMetadata[] = [];

  console.log('🚀 Starting comprehensive database backup...');
  console.log(`📁 Backup directory: ${backupDir}\n`);

  // Ensure backup directory exists
  if (!fs.existsSync(backupDir)) {
    fs.mkdirSync(backupDir, { recursive: true });
  }

  const timestamp = new Date().toISOString();

  try {
    // 1. Backup odoo_contacts
    backupMetadata.push(
      await backupTable(
        'odoo_contacts',
        () => prisma.odooContact.findMany({
          include: {
            matchedHospital: {
              include: {
                city: true,
                type: true,
                subtype: true
              }
            }
          }
        }),
        backupDir
      )
    );

    // 2. Backup hospitals
    backupMetadata.push(
      await backupTable(
        'hospitals',
        () => prisma.hospitalReference.findMany({
          include: {
            city: true,
            type: true,
            subtype: true,
            contacts: true,
            personContacts: true,
            contactPersons: true,
            odooContacts: true
          }
        }),
        backupDir
      )
    );

    // 3. Backup trello_matches
    backupMetadata.push(
      await backupTable(
        'trello_matches',
        () => prisma.trelloMatches.findMany({
          include: {
            hospital: {
              include: {
                city: true,
                type: true,
                subtype: true
              }
            }
          }
        }),
        backupDir
      )
    );

    // 4. Backup contacts
    backupMetadata.push(
      await backupTable(
        'contacts',
        () => prisma.contact.findMany({
          include: {
            jobTitle: true,
            hospital: {
              include: {
                city: true,
                type: true,
                subtype: true
              }
            }
          }
        }),
        backupDir
      )
    );

    // 5. Backup person_contacts
    backupMetadata.push(
      await backupTable(
        'person_contacts',
        () => prisma.personContact.findMany({
          include: {
            jobTitle: true,
            hospital: {
              include: {
                city: true,
                type: true,
                subtype: true
              }
            }
          }
        }),
        backupDir
      )
    );

    // 6. Backup cities
    backupMetadata.push(
      await backupTable(
        'cities',
        () => prisma.city.findMany({
          include: {
            hospitals: true
          }
        }),
        backupDir
      )
    );

    // 7. Backup job_titles
    backupMetadata.push(
      await backupTable(
        'job_titles',
        () => prisma.jobTitle.findMany({
          include: {
            personContacts: true,
            contactPersons: true
          }
        }),
        backupDir
      )
    );

    // 8. Backup hospital_types
    backupMetadata.push(
      await backupTable(
        'hospital_types',
        () => prisma.hospitalType.findMany({
          include: {
            subtypes: true,
            hospitals: true
          }
        }),
        backupDir
      )
    );

    // 9. Backup hospital_subtypes
    backupMetadata.push(
      await backupTable(
        'hospital_subtypes',
        () => prisma.hospitalSubtype.findMany({
          include: {
            parentType: true,
            hospitals: true
          }
        }),
        backupDir
      )
    );

    // 10. Backup trello_raw_datas
    backupMetadata.push(
      await backupTable(
        'trello_raw_datas',
        () => prisma.rawData.findMany(),
        backupDir
      )
    );

    // Get database version
    const dbVersion = await getDatabaseVersion();

    // Calculate totals
    const totalRecords = backupMetadata.reduce((sum, meta) => sum + meta.recordCount, 0);
    const totalSize = backupMetadata.reduce((sum, meta) => sum + (meta.fileSize || 0), 0);

    // Create summary
    const summary: BackupSummary = {
      backupTimestamp: timestamp,
      databaseVersion: dbVersion,
      tables: backupMetadata,
      totalRecords,
      totalSize
    };

    // Write summary file
    const summaryPath = path.join(backupDir, 'backup_summary.json');
    fs.writeFileSync(summaryPath, JSON.stringify(summary, null, 2), 'utf-8');

    // Print summary
    console.log('\n' + '='.repeat(80));
    console.log('✅ BACKUP COMPLETED SUCCESSFULLY');
    console.log('='.repeat(80));
    console.log(`\n📊 Summary:`);
    console.log(`   Timestamp: ${timestamp}`);
    console.log(`   Database Version: ${dbVersion.split('\n')[0]}`);
    console.log(`   Total Tables: ${backupMetadata.length}`);
    console.log(`   Total Records: ${totalRecords.toLocaleString()}`);
    console.log(`   Total Size: ${(totalSize / 1024 / 1024).toFixed(2)} MB`);
    console.log('\n📋 Table Details:');

    backupMetadata.forEach((meta) => {
      const sizeMB = ((meta.fileSize || 0) / 1024 / 1024).toFixed(2);
      console.log(`   - ${meta.tableName.padEnd(25)} : ${String(meta.recordCount).padStart(6)} records (${sizeMB} MB)`);
    });

    console.log(`\n📁 Backup location: ${backupDir}`);
    console.log(`📄 Summary file: backup_summary.json\n`);

  } catch (error) {
    console.error('\n❌ Backup failed:', error);
    throw error;
  } finally {
    await prisma.$disconnect();
  }
}

// Run the backup
createBackups()
  .then(() => {
    console.log('🎉 Backup process completed successfully!');
    process.exit(0);
  })
  .catch((error) => {
    console.error('💥 Backup process failed:', error);
    process.exit(1);
  });

File: backups/20251107_125734/create_json_backups.py
#!/usr/bin/env python3
"""
Comprehensive JSON backup script for orubacontacts database
Creates separate JSON files for each table with metadata
"""

import psycopg2
import json
import os
from datetime import datetime
from pathlib import Path
from typing import Dict, List, Any

# Database connection parameters
DB_CONFIG = {
    'host': 'localhost',
    'port': 5432,
    'database': 'orubacontacts',
    'user': 'postgres',
    'password': 'postgres'
}

# Backup directory
BACKUP_DIR = '/Users/fatihalkan/Documents/GitHub/orubacontacts/backups/20251107_125734'

# Tables to backup
TABLES = [
    'odoo_contacts',
    'hospitals',
    'trello_matches',
    'contacts',
    'person_contacts',
    'cities',
    'job_titles',
    'hospital_types',
    'hospital_subtypes',
    'trello_raw_datas'
]

def get_db_connection():
    """Create and return a database connection"""
    try:
        conn = psycopg2.connect(**DB_CONFIG)
        return conn
    except Exception as e:
        print(f"❌ Error connecting to database: {e}")
        raise

def get_database_version(conn) -> str:
    """Get PostgreSQL version"""
    try:
        with conn.cursor() as cur:
            cur.execute("SELECT version();")
            return cur.fetchone()[0]
    except Exception as e:
        print(f"⚠️  Could not get database version: {e}")
        return "Unknown"

def get_table_columns(conn, table_name: str) -> List[str]:
    """Get column names for a table"""
    with conn.cursor() as cur:
        cur.execute(f"""
            SELECT column_name
            FROM information_schema.columns
            WHERE table_name = '{table_name}'
            AND table_schema = 'public'
            ORDER BY ordinal_position;
        """)
        return [row[0] for row in cur.fetchall()]

def backup_table(conn, table_name: str) -> Dict[str, Any]:
    """Backup a single table to JSON"""
    print(f"\n📦 Backing up {table_name}...")

    try:
        # Get columns
        columns = get_table_columns(conn, table_name)

        # Fetch all records
        with conn.cursor() as cur:
            cur.execute(f"SELECT * FROM {table_name};")
            rows = cur.fetchall()

        # Convert to list of dicts
        records = []
        for row in rows:
            record = {}
            for i, column in enumerate(columns):
                value = row[i]
                # Handle datetime serialization
                if isinstance(value, datetime):
                    value = value.isoformat()
                record[column] = value
            records.append(record)

        record_count = len(records)
        print(f"   Found {record_count} records")

        # Create backup object
        backup = {
            'metadata': {
                'tableName': table_name,
                'recordCount': record_count,
                'timestamp': datetime.now().isoformat(),
                'backupVersion': '1.0.0',
                'columns': columns
            },
            'data': records
        }

        # Write to file
        file_name = f"{table_name}_backup.json"
        file_path = os.path.join(BACKUP_DIR, file_name)

        with open(file_path, 'w', encoding='utf-8') as f:
            json.dump(backup, f, indent=2, ensure_ascii=False)

        # Get file size
        file_size = os.path.getsize(file_path)
        file_size_mb = file_size / (1024 * 1024)

        print(f"   ✅ Saved to {file_name} ({file_size_mb:.2f} MB)")

        return {
            'tableName': table_name,
            'recordCount': record_count,
            'timestamp': backup['metadata']['timestamp'],
            'filePath': file_path,
            'fileSize': file_size
        }

    except Exception as e:
        print(f"   ❌ Error backing up {table_name}: {e}")
        raise

def create_backups():
    """Main backup function"""
    print("🚀 Starting comprehensive database backup...")
    print(f"📁 Backup directory: {BACKUP_DIR}\n")

    # Ensure backup directory exists
    Path(BACKUP_DIR).mkdir(parents=True, exist_ok=True)

    backup_metadata = []
    conn = None

    try:
        # Connect to database
        print("🔌 Connecting to database...")
        conn = get_db_connection()
        print("✅ Connection successful\n")

        # Get database version
        db_version = get_database_version(conn)

        # Backup each table
        for table in TABLES:
            metadata = backup_table(conn, table)
            backup_metadata.append(metadata)

        # Calculate totals
        total_records = sum(meta['recordCount'] for meta in backup_metadata)
        total_size = sum(meta['fileSize'] for meta in backup_metadata)

        # Create summary
        summary = {
            'backupTimestamp': datetime.now().isoformat(),
            'databaseVersion': db_version,
            'tables': backup_metadata,
            'totalRecords': total_records,
            'totalSize': total_size
        }

        # Write summary file
        summary_path = os.path.join(BACKUP_DIR, 'backup_summary.json')
        with open(summary_path, 'w', encoding='utf-8') as f:
            json.dump(summary, f, indent=2)

        # Print summary
        print("\n" + "=" * 80)
        print("✅ BACKUP COMPLETED SUCCESSFULLY")
        print("=" * 80)
        print(f"\n📊 Summary:")
        print(f"   Timestamp: {summary['backupTimestamp']}")
        print(f"   Database Version: {db_version.split(chr(10))[0]}")
        print(f"   Total Tables: {len(backup_metadata)}")
        print(f"   Total Records: {total_records:,}")
        print(f"   Total Size: {total_size / (1024 * 1024):.2f} MB")
        print("\n📋 Table Details:")

        for meta in backup_metadata:
            size_mb = meta['fileSize'] / (1024 * 1024)
            print(f"   - {meta['tableName']:<25} : {meta['recordCount']:>6,} records ({size_mb:.2f} MB)")

        print(f"\n📁 Backup location: {BACKUP_DIR}")
        print(f"📄 Summary file: backup_summary.json\n")

    except Exception as e:
        print(f"\n❌ Backup failed: {e}")
        raise
    finally:
        if conn:
            conn.close()
            print("🔌 Database connection closed")

if __name__ == '__main__':
    try:
        create_backups()
        print("🎉 Backup process completed successfully!")
        exit(0)
    except Exception as e:
        print(f"💥 Backup process failed: {e}")
        exit(1)
